src/Bedrock.Framework/Protocols/MessagePipeReader.cs

```diff
@@ -0,0 +1,224 @@
+        {
+            if (_isThisCompleted)
+            {
+                ThrowReadAfterCompleted();
+            }
+
+            _allExamined = _message.Slice(examined).IsEmpty;
+
+            var consumedLength = (int)_message.Slice(_message.Start, consumed).Length;
+
+            if (_backlog.UnconsumedWrittenCount > 0)
+            {
+                _backlog.Consume(consumedLength);
+            }
+            else
+            {
+                var unconsumed = _message.Slice(consumed);
+                foreach (var m in unconsumed)
+                {
+                    _backlog.Write(m.Span);
+                }
+            }
+
+            // We cannot advance the underlying reader twice without calling _reader.ReadAsync/TryRead in between
+            // If _isCompleted is true we never call reader.ReadAsync/TryRead
+            // So we cannot call _reader.AdvanceTo, and there is no need to because either the buffer was empty,
+            // or we called AdvanceTo in TryCreateReadResult
+            if (!_isCompleted)
+            {
+                // REVIEW: Use the correct value for examined
+                _reader.AdvanceTo(_consumed, _examined);
+            }
+
+            _advanced = true;
+        }
+
+        public override void CancelPendingRead()
+        {
+            _reader.CancelPendingRead();
+        }
+
+        public override void Complete(Exception exception = null)
+        {
+            _isThisCompleted = true;
+            _reader.Complete();
+        }
+
+        public override async ValueTask<ReadResult> ReadAsync(CancellationToken cancellationToken = default)
+        {
+            if (_isThisCompleted)
+            {
+                ThrowReadAfterCompleted();
+            }
+
+            if (_isCompleted)
+            {
+                _message = new ReadOnlySequence<byte>(_backlog.WrittenMemory);
+                return new ReadResult(_message, false, isCompleted: true);
+            }
+
+
+            if (!_advanced)
+            {
+                AdvanceTo(_message.Start);
+            }
+
+            while (true)
+            {
+                var result = await _reader.ReadAsync(cancellationToken).ConfigureAwait(false);
+
+                if (TryCreateReadResult(result, out var readResult))
+                {
+                    return readResult;
+                }
+            }
+        }
+
+        public override bool TryRead(out ReadResult readResult)
+        {
+            if (_isThisCompleted)
+            {
+                ThrowReadAfterCompleted();
+            }
+
+            if (_isCompleted)
+            {
+                if (_allExamined)
+                {
+                    readResult = default;
+                    return false;
+                }
+
+                _message = new ReadOnlySequence<byte>(_backlog.WrittenMemory);
+                readResult = new ReadResult(_message, false, isCompleted: true);
```

Comment:
We'll return false for isCancelled here even if the user calls CancelPendingRead before the call.
I don't think this is a problem, given that we return immediately.